stl/inc/variant

```diff
@@ -1460,29 +1466,38 @@ struct _Variant_dispatcher;
 
 template <size_t... _Is>
 struct _Variant_dispatcher<index_sequence<_Is...>> {
-    template <class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
-    _NODISCARD static constexpr _Variant_visit_result_t<_Callable, _Types...> _Dispatch(
-        _Callable&& _Obj, _Types&&... _Args) {
+    template <class _Ret, class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>

```

Comment:
Given that you change the world anyway would it make sense to use `std::disjunction` here as that short circuits? 

```cpp
bool _Any_valueless = disjunction_v<_Is == 0), ...>
```

Reply:
If I'm not mistaken, I do believe fold expression already short circuit.

Reply:
As I understand it that is not the case and especially mentioned as a use case of {con, dis}junction:

https://en.cppreference.com/w/cpp/types/disjunction

Reply:
From the notes

> The short-circuit instantiation differentiates disjunction from fold expressions: a fold expression like (... || Bs::value) instantiates every B in Bs, while std::disjunction_v<Bs...> stops instantiation once the value can be determined.

Reply:
Yes, when you use a fold expression such as `((_Is == 0) || ...)` every `_Is` is instantiated, however, when they they are being evaluated, short circuiting will still occur.
e.g.
```
template<std::size_t... Is>
constexpr bool b = ((Is == 0) || ...);

bool result = b<1, 0, 2>;
``` 
The fold expression will be instantiated such as
```
constexpr bool b = (1 == 0 || 0 == 0 || 2 == 0);
```
So each `Is` is instantiated, however, once `0 == 0` is evaluated as true, `2 == 0` will not be evaluated.
On the other hand, `std::disjunction` will only instantiate each value as needed.
**tldr:** Both will short circuit, it's just that the fold expression will instantiate every value.
Regardless, I do also prefer `std::disjunction` and will probably change it (thanks for the help :D).

Reply:
If the `_Is` were expensive to evaluate, then `disjunction` would make sense here. Given that the `_Is` are actually integer constants here, I doubt we can beat the compile time of this simple fold expression.